File: apps/proxmox-vm-worker/vmTypes/service.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, field_validator
from uuid import UUID
from ulid import ULID
import json


def generate_ulid() -> str:
    # For illustration purposes, we use uuid4. Replace with a ULID generator if needed.
    return ULID().str


class Metadata(BaseModel):
    initial_sku: str
    initial_price: int


class Service(BaseModel):
    id: str = Field(default_factory=generate_ulid)
    user_id: UUID
    service_type: str
    hostname: str
    template_id: str
    os_name: str
    os_version: str
    public_key_id: str
    username: str
    metadata: Metadata
    sku_id: str
    current_sku_id: str
    initial_sku_id: str
    subscription_active: bool = False
    subscription_id: Optional[str] = None
    initial_checkout_id: Optional[str] = None
    status: str
    status_reason: Optional[str] = None
    payment_ids: List[str]
    payment_status: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    deleted_at: Optional[datetime] = None
    accountId: Optional[str] = Field(None)
    proxmox_node: Optional[str] = None
    proxmox_vm_id: Optional[str] = None

    @field_validator("metadata", mode="before")
    def parse_metadata(cls, value):
        if isinstance(value, str):
            try:
                parsed = json.loads(value)
            except Exception as e:
                raise ValueError("Invalid JSON for metadata") from e
            # Handle double-encoded JSON strings
            if isinstance(parsed, str):
                try:
                    parsed = json.loads(parsed)
                except Exception as e:
                    raise ValueError("Invalid nested JSON for metadata") from e
            if isinstance(parsed, dict):
                return Metadata(**parsed)
            else:
                raise ValueError("metadata JSON must be a mapping")
        if isinstance(value, dict):
            return Metadata(**value)
        return value

    model_config = {
        "json_encoders": {
            ULID: lambda v: str(v),
            datetime: lambda dt: dt.isoformat(),
        }
    }

File: internal/types/service.go
// /home/xn4p4lm/projects/andromeda/gigabit-host/internal/types/vmObject/service.go
package types

import (
	"github.com/google/uuid"
	"github.com/oklog/ulid/v2"
)

type ServiceMetadata struct {
	InitialSku  string `json:"initial_sku"`
	InitalPrice string `json:"initial_price"`
}

type Service struct {
	ID                 ulid.ULID       `json:"id" db:"id"`
	UserID             uuid.UUID       `json:"user_id" db:"user_id"`
	ServiceType        string          `json:"service_type" db:"service_type"`
	Hostname           string          `json:"hostname" db:"hostname"`
	TemplateID         string          `json:"template_id" db:"template_id"`
	OSName             string          `json:"os_name" db:"os_name"`
	OSVersion          string          `json:"os_version" db:"os_version"`
	PublicKeyID        string          `json:"public_key_id" db:"public_key_id"`
	Username           string          `json:"username" db:"username"`
	Metadata           ServiceMetadata `json:"metadata" db:"metadata"`
	SKUID              string          `json:"sku_id" db:"sku_id"`
	CurrentSKUID       string          `json:"current_sku_id" db:"current_sku_id"`
	InitialSKUID       string          `json:"initial_sku_id" db:"initial_sku_id"`
	SubscriptionActive bool            `json:"subscription_active" db:"subscription_active"`
	SubscriptionID     *string         `json:"subscription_id,omitempty" db:"subscription_id"`
	InitialCheckoutID  *string         `json:"initial_checkout_id,omitempty" db:"initial_checkout_id"`
	Status             string          `json:"status" db:"status"`
	StatusReason       *string         `json:"status_reason,omitempty" db:"status_reason"`
	PaymentIDs         []string        `json:"payment_ids" db:"payment_ids"`
	PaymentStatus      *string         `json:"payment_status,omitempty" db:"payment_status"`
	CreatedAt          string          `json:"created_at" db:"created_at"`
	UpdatedAt          string          `json:"updated_at" db:"updated_at"`
	DeletedAt          *string         `json:"deleted_at,omitempty" db:"deleted_at"`
	AccountID          *string         `json:"account_id,omitempty" db:"account_id"`
	ProxmoxNode        *string         `json:"proxmox_node,omitempty" db:"proxmox_node"`
	ProxmoxVMID        *string         `json:"proxmox_vm_id,omitempty" db:"proxmox_vm_id"`
}
